Parallel Environment Execution

Bam4d opened the pull request:
Taking inspiration from stable-baselines' vector implementation, but with some additional optimizations... the ParallelListEnv takes a `num_processes` and `num_envs` parameter. and will split the envs into equal proportions on each process.

This is better than having a process per-environment as it reduces the latency/observations ratio. (having 10 environments per process, means we can get 10 observations with only 1 send/receive from the process).

Setting `num_processes==num_envs` is equivalent to SubProcVecEnv. This has a overhead of 1 observation per send/receive.

Setting `num_processes==1` means you have no parallelization.

Setting `num_processes==8` and `num_envs=64` will give you 8 envs per-process. This has a reduced overhead of 8 observations per send/recieve.

With Griddly I can achieve about 6K SPS, up from previous 3K SPS, when the envs are split across 8 processes.
![image](https://user-images.githubusercontent.com/1370765/146036471-3e7d24dd-6151-468e-8a55-43bf2a8eed9e.png)

cswinter commented:
WOOT! Love seeing those performance gains!

Haven't looked at the code super closely yet, some general thoughts:
- Do we need to split the `VecEnv` into `VecEnv`/`BatchEnv`? I feel like we should have just the `BatchEnv` but I might be missing the use case.
- Let's push the `batch_obs` into the processes so it's parallelized as well, could give another nice performance boost especially if there are many environments per process.
- Can we avoid using (cloud)pickle? Not a high priority if it's difficult to and the current implementation is likely fine, but pickle is insecure, and having code that accepts data over a network connection and then feeds it into pickle makes me feel somewhat uneasy. The obs batch is quite structured so we should be able to use message pack (there's an example in enn_ppo/sample_recorder.py). To serialize a RaggedBuffer, we can turn it into numpy arrays of flattened data + sequence lengths. One potential downside is that now environments can't stuff arbitrary extra Python objects into observation and things won't just work if they do, though currently that's not supported either way.

Bam4d commented:
So the reason for splitting ObsBatch, VecEnv, BatchEnv is so that the communication layer i.e cloudpickle can serialize and send across whatever protocol its using. cloudpickle could not serialize ObsBatch because of the `RaggedBuffer`

If we can easily serialize the `RaggedBuffer` stuff then I think switching over to msg_pack and removing cloudpickle + pushing batch_obs into the parallelization would be possible.

I'll see if I can do the RaggedBuffer serialization + messagepack.

>One potential downside is that now environments can't stuff arbitrary extra Python objects into observation and things won't just work if they do, though currently that's not supported either way.

Hmm this might be a bit painful for future use-cases that need information from the environment we are not thinking about. (having these APIs as flexible as possible is ideal because we don't know what people might try to do with it in the future)... I'm sure there's a smart way to flatten dict data with something like messagepack. I'll look into it.

cswinter commented:
Yeah dicts/primitive types/numpy arrays works fine with message pack, but e.g. dataclasses and Python objects won't. I'm not sure there's a great solution here, maybe what we should do is use messagepack by default for now, give a really good error message if messagepack serialization fails, and add a way to explicitly opt-in to pickle instead if it turns out to be necessary.

Converting ragged buffer from/into numpy arrays is fairly straightforward:

```python
>>> import ragged_buffer
>>> from ragged_buffer import RaggedBufferF32
>>> import numpy as np
>>> flattened = np.array(
...     [
...         [2.0, 3.0, 1.0],
...         [1.0, 2.0, 3.0],
...         [2.0, 3.0, 1.0],
...         [1.0, 2.0, 3.0],
...         [1.4, 2.4, 3.4],
...         [2.0, 3.0, 1.0],
...         [1.0, 2.0, 3.0],
...         [1.4, 2.4, 3.4],
...         [1.4, 2.4, 3.4],
...     ],
...     dtype=np.float32,
... )
>>> lenghts = np.array([2, 3, 4, 0], dtype=np.int64)
>>> rb = RaggedBufferF32.from_flattened(
...     flattened=flattened, lengths=lenghts,
... )
>>> flattened2 = rb.as_array()
>>> lenghts2 = rb.size1()
>>> np.all(lenghts == lenghts2)
True
>>> np.all(flattened == flattened2)
True
```

Bam4d commented:
@cswinter I managed to squeeze anoter 1K SPS by sellotaping msgpack as a replacement for cloudpickle (and shimming some othet tricks so it can actually serialize the objects we are using) 
![image](https://user-images.githubusercontent.com/1370765/146236396-5b69050a-0cc3-4aeb-a7f7-3077941d7976.png)


HOWEVER: I'm concerned that I might be introducing a bug/there is already a bug in the way that ObsBatch is being built/merged.

if we have a List[Observation] of 100 observations ... and in some of those observations we are missing entities taht have been seen before... we are not pushing 0-length ragged arrays into the batch..

In this bit of code here we are only adding to `entities[k]` if it exists in the observation `o`... but if its not in that observation we are not adding a 0-length feature vector for that particular entity... meaning future entities are getting added to the incorrect batch.
https://github.com/entity-neural-network/incubator/blob/main/entity_gym/entity_gym/environment.py#L200

I'm not 100% sure that I'm understanding exactly how the ragged buffer / batch_obs is working so this might be me misunderstanding whats going on here...

Either way, let me know what you think!

cswinter commented:
Awesome!
Good catch on the empty observation. Maybe what should do there is iterate over the obs_space so we're guaranteed the same set of entities on each observation. If an entity is missing, we can either error or default to 0 entities of that type.

Bam4d commented:
> Awesome! Good catch on the empty observation. Maybe what should do there is iterate over the obs_space so we're guaranteed the same set of entities on each observation. If an entity is missing, we can either error or default to 0 entities of that type.

Glad to know I wasn't totally bamboozled and it was a bug :)

Yes i thought about iterating over the obs_space. I will implement that tomorrow :D

Bam4d commented:
@cswinter I think this is good for another review pass. I fixed the batching and merging operations and added a set of tests to make sure they stay consistent.

I have some Griddly stuff to work on for a bit (big release with lots more examples and documentation planned), I'll pick up more of this work in the new year.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -184,7 +203,67 @@ class ObsBatch:
+    entities = {}
+    ids: Sequence[Sequence[EntityID]] = []
+    action_masks: Dict[
+        str, Union[CategoricalActionMaskBatch, SelectEntityActionMaskBatch]
```
We could add a type alias `ActionMaskBatch = Union[CategoricalActionMaskBatch, SelectEntityActionMaskBatch]`

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -184,7 +203,67 @@ class ObsBatch:
+    reward = np.concatenate((a.reward, b.reward))
+    done = np.concatenate((a.done, b.done))
+
+    for k, v in a.end_of_episode_info.items():  # type: ignore
```
Why do we need to ignore the type here?

cswinter replied:
I think we can use the dict `update` method.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -184,7 +203,67 @@ class ObsBatch:
+        entities[k].extend(b.entities[k])
+
+    # merge ids
+    ids.extend(a.ids)  # type: ignore
```
There ought to be a way of making the types work out.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -184,7 +203,67 @@ class ObsBatch:
 
 
-def batch_obs(obs: List[Observation]) -> ObsBatch:
+def merge_obs(a: Optional[ObsBatch], b: ObsBatch) -> ObsBatch:
```
I'd make both of the input args non-optional and check for none outside of the function.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -184,7 +203,67 @@ class ObsBatch:
+    # merge masks
+    for k in a.action_masks.keys():
+        action_masks[k] = a.action_masks[k]
+        action_masks[k].extend(b.action_masks[k])
```
This mutates `a`. We should either create a copy of `action_masks[k]` here, or change everything to just mutate `a` and make that more obvious in the type signature (by returning `None`).

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -314,6 +416,9 @@ def reset(self, obs_filter: ObsSpace) -> Observation:
     def act(self, action: Mapping[str, Action], obs_filter: ObsSpace) -> Observation:
         return self.__class__.filter_obs(self._act(action), obs_filter)
 
+    def close(self) -> None:
```
This should probably just do nothing by default rather than error, or be an `@abstractmethod`.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -388,4 +506,205 @@ def act(
+                return RaggedBufferI64.from_flattened(flattened, lengths)
+        elif "__classname__" in obj:
+            cls_name = globals()[obj["__classname__"]]
+            return cls_name(**obj["data"])
```
Why do we need this? I assume to serialize dataclasses? This probably allows for arbitrary code execution, if we need this we should probably at least have a whitelist of allowed classes.

Bam4d replied:
Yes its so we can serialize classes like ObsBatch etc. I'm not sure if it does allow for arbitrary execution. My understanding is that `globals()` only allows classes that are in scope, so the Whitelist is essentially already included in the imports of the file.

cswinter replied:
Well, it would be very easy to accidentally add an import that allows for something destructive, so I'd sleep better with an explicit whitelist. Globals has access to things like `__builtins__` and `__loader__`, which I've not found an obvious way to utilize, but I wouldn't be completely certain that it's impossible to get access to something like `eval`, particularly over multiple iterations.

```
>>> globals()
{'__name__': '__main__', '__doc__': None, '__package__': None, '__loader__': <class '_frozen_importlib.BuiltinImporter'>, '__spec__': None, '__annotations__': {}, '__builtins__': <module 'builtins' (built-in)>}
```

Bam4d replied:
Sounds good, I agree its better to be defensive here. I'm no expert on python security, so whitelisting sounds smart.

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -388,4 +506,205 @@ def act(
+        try:
+            cmd, data = remote.recv()
+            if cmd == "act":
+                # def act(self, action: Mapping[str, Action], obs_filter: ObsSpace) -> Observation:
```
remove

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -388,4 +506,205 @@ def act(
+                observation = envs.act(data[0], data[1])
+                remote.send(observation)
+            elif cmd == "reset":
+                # def reset(self, obs_filter: ObsSpace) -> Observation:
```
remove

cswinter reviewed (commented):
Mostly LGTM, just a few more details to fix (whenever you have time). Good luck with the Griddly release!

cswinter commented on entity_gym/entity_gym/environment.py:
```diff
@@ -174,57 +183,136 @@ def clear(self) -> None:
+    merges two obs_batches
+    """
+
+    entities = a.entities.copy()
```
This creates a shallow copy, so we still end up mutating the original list. Easiest approach could be to just make the method mutate the first arg, if we make this a method on `ObsBatch` that returns `None` it would be fairly obvious that that's what's happening.

cswinter reviewed (approved):
LGTM, I just have one more nit about `merge_obs`.

Bam4d commented:
@cswinter Added merge_obs as a method on ObsBatch. Will merge now.
